### PR TITLE
Initial Length Harmonies

### PR DESCRIPTION
**Pull request by BSalita**
Initial commits for harmonizing lengths (Index, size_t, uint32_t ...)

### CODE REVIEW

File: src/emscripten-optimizer/simple_ast.h [hunk 1]
```diff
@@ -920,22 +920,22 @@ struct JSPrinter {
         char *test = end;
         // remove zeros, and also doubles can use at most 24 digits, we can truncate any extras even if not zero
         while ((*test == '0' || test - buffer > 24) && test > buffer) test--;
-        int num = end - test;
+        size_t num = end - test;
```
Comment (dschuff) [inline src/emscripten-optimizer/simple_ast.h]: Technically this ought to be `ptrdiff_t`, a signed quantity.

File: src/emscripten-optimizer/simple_ast.h [hunk 2]
```diff
@@ -920,22 +920,22 @@ struct JSPrinter {
...
         if (num >= 3) {
           test++;
           test[0] = 'e';
           if (num < 10) {
-            test[1] = '0' + num;
+            test[1] = (char)('0' + num);
```
Comment (dschuff) [inline src/emscripten-optimizer/simple_ast.h]: Maybe instead of casting everywhere here we should just introduce a char temp (convert num to a char once), and add that to `'0'`. We could also add an assert at the conversion time.

File: src/emscripten-optimizer/simple_ast.h (hunk 1 above)
Comment (BSalita) [inline src/emscripten-optimizer/simple_ast.h]: Absolutely. I'll change it and look for other instances.

File: src/support/archive.cpp [hunk 3]
```diff
@@ -193,7 +193,7 @@ Archive::child_iterator Archive::child_end() const { return Child(); }
 namespace {
 struct Symbol {
   uint32_t symbolIndex;
-  uint32_t stringIndex;
+  ptrdiff_t stringIndex;
```
Comment (dschuff) [inline src/support/archive.cpp]: Actually looking at the code, this use of stringIndex/strchr is straight-up wrong. It doesn't matter right now because we don't use archives with symbol tables yet. You can leave this as-is (it will warn, as it should... because it's wrong :))

File: src/support/archive.cpp [hunk 4]
```diff
@@ -210,8 +210,8 @@ static uint32_t read32be(const uint8_t* buf) {
 }
 
 void Archive::dump() const {
-  printf("Archive data %p len %zu, firstRegularData %p\n", data.data(),
-         data.size(), firstRegularData);
+	printf("Archive data %p len %zu, firstRegularData %p\n", data.data(),
```
Comment (dschuff) [inline src/support/archive.cpp]: extra indent?

File: src/wasm.h [hunk 5]
```diff
@@ -90,6 +90,9 @@ struct Name : public cashew::IString {
 
 // An index in a wasm module
 typedef uint32_t Index;
+inline Index toIndex(size_t index) {
```
Comment (dschuff) [inline src/wasm.h]: Is the plan to use this in more places? If not, let's just put a cast where the one use is. If so (and especially if there will be a lot) it might be cleaner to do something with an implicit conversion like we did with Address. (that can be a different PR).

**Comment by dschuff** [comment]
Thanks for taking this on!

File: src/support/archive.cpp (hunk 4 above)
Comment (BSalita) [inline src/support/archive.cpp]: Visual Studio editor artifact. I'll have to change Visual Studio indentation settings. Apparently there's no pretty print scheduled? So it's all up to the programmer to manually pretty print?

File: src/emscripten-optimizer/simple_ast.h (hunk 2 above)
Comment (BSalita) [inline src/emscripten-optimizer/simple_ast.h]: I'm not sure what you mean. The issue is that addition operators are promoting to int.

File: src/wasm.h (hunk 5 above)
Comment (BSalita) [inline src/wasm.h]: Three files, so far: Print.cpp, wasm-binary.h, wasm-builder.h

**Comment by BSalita** [comment]
Thanks for putting up with me. I'm a bit clumsy with the tools and teamwork. I'll gradually improve.
I should be able to actively participate for another month, maybe much longer.

    On Friday, May 13, 2016 5:56 PM, Derek Schuff <notifications@github.com> wrote:
 

 Thanks for taking this on!—

File: src/emscripten-optimizer/simple_ast.h (hunk 2 above)
Comment (dschuff) [inline src/emscripten-optimizer/simple_ast.h]: Right; `'0' + num` promotes to int because `num` is an int. Actually what I had in mind wouldn't really work. I think let's just put the casts on `num` expressions directly rather than the promoted result, e.g. `test[1] = '0' + (char)(num / 10);` That makes it clearer that there won't be overflow (since we are testing the value of `num`) and prevents the promotion altogether.

File: src/support/archive.cpp (hunk 4 above)
Comment (dschuff) [inline src/support/archive.cpp]: We had talked before about making a `.clang-format` file but I think it hasn't been checked in yet. Also the files in Binaryen now come from several sources, so the formatting isn't in good shape at all right now... :(

**Comment by BSalita** [comment]
Looks like I need to install gcc and clang so I can provide better quality PRs. Visual Studio alone isn't sufficient. Fortunately, I can now use gcc and clang in Microsoft's new bash shell feature. 

    On Friday, May 13, 2016 6:48 PM, Bob Salita <bsalita@yahoo.com> wrote:
 

 Thanks for putting up with me. I'm a bit clumsy with the tools and teamwork. I'll gradually improve.
I should be able to actively participate for another month, maybe much longer.

    On Friday, May 13, 2016 5:56 PM, Derek Schuff <notifications@github.com> wrote:
 

 Thanks for taking this on!—

File: src/wasm.h (hunk 5 above)
Comment (dschuff) [inline src/wasm.h]: OK; let's maybe handle `Index` in another PR then.

**Comment by dschuff** [comment]
You don't necessarily need to install those. You can always rely on Travis to see what happens with the Linux build. (There's also clang for Windows; I've never tried that though).

**Comment by BSalita** [comment]
The benefit to having cleaner PRs is less time teammates look at my code and more time for their own work.

    On Friday, May 13, 2016 6:55 PM, Derek Schuff <notifications@github.com> wrote:
 

 You don't necessarily need to install those. You can always rely on Travis to see what happens with the Linux build. (There's also clang for Windows; I've never tried that though).—

File: src/emscripten-optimizer/simple_ast.h (hunk 2 above)
Comment (BSalita) [inline src/emscripten-optimizer/simple_ast.h]: I believe '0' is also an int.

Comment (dschuff) [inline src/emscripten-optimizer/simple_ast.h]: It is in C; It's actually not in C++ (http://en.cppreference.com/w/cpp/language/character_literal)

Comment (BSalita) [inline src/emscripten-optimizer/simple_ast.h]: Yet another thing I learned today.

    On Friday, May 13, 2016 10:38 PM, Derek Schuff <notifications@github.com> wrote:
 

 In src/emscripten-optimizer/simple_ast.h:>          if (num >= 3) {
>            test++;
>            test[0] = 'e';
>            if (num < 10) {
> -            test[1] = '0' + num;
> +            test[1] = (char)('0' + num);
It is in C; It's actually not in C++ (http://en.cppreference.com/w/cpp/language/character_literal)—

**Comment by BSalita** [comment]
Is there some action I have to take to get this PR merged?

File: src/support/archive.cpp [hunk 6]
```diff
@@ -211,7 +211,7 @@ static uint32_t read32be(const uint8_t* buf) {
 
 void Archive::dump() const {
   printf("Archive data %p len %zu, firstRegularData %p\n", data.data(),
-         data.size(), firstRegularData);
```
Comment (kripken) [inline src/support/archive.cpp]: what changed here?

File: src/wasm.h [hunk 7]
```diff
@@ -90,6 +90,9 @@ struct Name : public cashew::IString {
 
 // An index in a wasm module
 typedef uint32_t Index;
```
Comment (kripken) [inline src/wasm.h]: newline after this typedef please

File: src/wasm.h (hunk 5 above)
Comment (kripken) [inline src/wasm.h]: Method sounds good to me since there are going to be multiple uses. Can leave the others for a followup PR, I agree.

How about asserting on the right range in the function?

File: src/support/archive.cpp (hunk 6 above)
Comment (BSalita) [inline src/support/archive.cpp]: Just a whitespace change.

File: src/wasm.h (hunk 5 above)
Comment (BSalita) [inline src/wasm.h]: Do you mean add an assert() within toIndex()? Yes, if max (size) is passed as an additional parameter, an assert could be added. Should I do that?

Comment (kripken) [inline src/wasm.h]: Yes, let's add an assert here. But I think its enough to check on the maximum size of the type (a more specific check would require some overhead in getting that information to here).

File: src/wasm.h (hunk 7 above)
Comment (BSalita) [inline src/wasm.h]: @kripken Assert added in toIndex(). I'm requesting a merge now. After merge, I'll go back and look over the additional suggestions. Changes are plentiful enough that I don't want to complicate the PR further.

Comment (BSalita) [inline src/wasm.h]: In wasm-binary.h, in mapLocals() line 586, I added a toIndex(). Looks like it will always assert. However, travis passes so I'm thinking the code is never executed. I can't tell if curr is suppose to be equal to size() correctly, or if curr is off by 1, or it's just a work in progress. Can you have a look?

Comment (kripken) [inline src/wasm.h]: The mapLocals code? It should definitely be executed, unless something very strange has gone wrong. Maybe add a printf to verify it is run?

The code looks right to me, `curr` values there begin at 0 and increase. Unless I'm missing a bug?

File: src/wasm.h [hunk 8]
```diff
@@ -91,6 +91,11 @@ struct Name : public cashew::IString {
...
 
+inline Index toIndex(size_t index, size_t max) {
+	//assert(index < max);
```
Comment (kripken) [inline src/wasm.h]: indentation looks wrong here

File: src/wasm.h [hunk 9]
```diff
@@ -91,6 +91,11 @@ struct Name : public cashew::IString {
 // An index in a wasm module
 typedef uint32_t Index;
 
+inline Index toIndex(size_t index, size_t max) {
```
Comment (kripken) [inline src/wasm.h]: I'm not sure about the second parameter. It does help in some cases, but we may not always have that information around, and we'd need to write a lot of `getFunction()->getNumLocals()`. I suggest just asserting on just the size of the type, `assert(index <= std::numeric_limits<uint32_t>::max())`.

Comment (BSalita) [inline src/wasm.h]: How about if I create an overload with one parameter that includes your suggested assert?
 assert(index <= std::numeric_limits<uint32_t>::max())
Or should I just change two to one parameter? 

    On Thursday, May 19, 2016 8:04 PM, Alon Zakai <notifications@github.com> wrote:
 

 In src/wasm.h:> @@ -91,6 +91,11 @@ struct Name : public cashew::IString {
>  // An index in a wasm module
>  typedef uint32_t Index;
>  
> +inline Index toIndex(size_t index, size_t max) {
I'm not sure about the second parameter. It does help in some cases, but we may not always have that information around, and we'd need to write a lot of getFunction()->getNumLocals(). I suggest just asserting on just the size of the type, assert(index <= std::numeric_limits<uint32_t>::max()).—

File: src/wasm.h (hunk 8 above)
Comment (BSalita) [inline src/wasm.h]: I'm being stupid. I removed the assert to test and then forgot about it. This explains why mapLocals didn't assert. I'll uncomment.
 

    On Thursday, May 19, 2016 8:01 PM, Alon Zakai <notifications@github.com> wrote:
 

 In src/wasm.h:> @@ -91,6 +91,11 @@ struct Name : public cashew::IString {
>  // An index in a wasm module
>  typedef uint32_t Index;
>  
> +inline Index toIndex(size_t index, size_t max) {
> +	//assert(index < max);
indentation looks wrong here—

File: src/wasm.h (hunk 9 above)
Comment (kripken) [inline src/wasm.h]: Let's do just one parameter for now. For better checks, I think we might find a better option, like a method on the function `func->toLocalIndex(i)`, for example, but let's leave that discussion for later.

File: src/wasm.h (hunk 8 above)
Comment (BSalita) [inline src/wasm.h]: I removed the second parameter of toIndex() and inserted the new assert. How's it looking now for a merge?

File: src/passes/Print.cpp [hunk 10]
```diff
@@ -454,7 +454,7 @@ struct PrintSExpression : public Visitor<PrintSExpression> {
     if (curr->params.size() > 0) {
       for (size_t i = 0; i < curr->params.size(); i++) {
```
Comment (kripken) [inline src/passes/Print.cpp]: instead of calling toIndex, we can just use Index as the type of i. I suppose then we'd need toIndex on the size() output? that is still less code clutter though.

but I would like to not even have the toIndex on the size() call. The array type could be modified, perhaps?

File: src/wasm-binary.h [hunk 11]
```diff
@@ -116,6 +116,7 @@ struct LEB {
   }
 };
 
+#define SIZE32LEB(v) U32LEB((uint32_t)(v))
```
Comment (kripken) [inline src/wasm-binary.h]: not sure I like a macro here.

maybe the bigger question is, do we need this cast? and the next cast after it, `int32_t ret = (int32_t)o.size();` - are those strictly necessary? they add code clutter and confusion. i'm not sure i understand why we need them.

File: src/wasm-binary.h [hunk 12]
```diff
@@ -1180,7 +1181,8 @@ class WasmBinaryBuilder {
     // read sections until the end
     while (more()) {
       auto nameSize = getU32LEB();
-      uint32_t sectionSize, before;
+	  uint32_t sectionSize;
```
Comment (kripken) [inline src/wasm-binary.h]: indentation looks broken here

**Comment by kripken** [comment]
I think the noncontroversial parts of this PR are where we change to use the right type, and do not add a cast or a macro. That makes the code strictly shorter and more correct. Might make sense to land that stuff first, and leave the rest for a followup? I'm ok either way.

File: src/passes/Print.cpp (hunk 10 above)
Comment (BSalita) [inline src/passes/Print.cpp]: The usual`for` idiom has array indexing within it. In this case there is none. So I could go either way. The toIndex() is needed somewhere because in 64-bit, sizeof(size_t) > sizeof(Index).

File: src/wasm-binary.h (hunk 11 above)
Comment (BSalita) [inline src/wasm-binary.h]: All `.size()` are size_t so a cast is needed for any 32-bit value. The reason for SIZE32LEB exists is to cast 64-bit size_t 32-bit LEB. I'm suspect that some of the int32_t in this file should be uint32_t or Index or some new name. I don't want to even peek at this issue until this PR is closed.

File: src/wasm-binary.h (hunk 12 above)
Comment (BSalita) [inline src/wasm-binary.h]: fixed

File: src/passes/Print.cpp (hunk 10 above)
Comment (kripken) [inline src/passes/Print.cpp]: There is another option though, of `params` being a type that has `size()` returning `Index`, instead of an `std::vector`. (Perhaps an `IndexedVector`?) Anyhow, just a thought, maybe it's a bad idea, but I also think lots of toIndex calls is bad as well.

The only think toIndex saves us from is if the number of params is more than can fit in a 32-bit value? I don't think that risk is worth the cast. I *do* think the cast is useful when receiving outside input, like when parsing s-expressions. But internally, like here, it feels like just clutter.

File: src/wasm-binary.h (hunk 11 above)
Comment (kripken) [inline src/wasm-binary.h]: I agree on not doing more in this PR. Perhaps let's take out the SIZE32LEB then, and later think more on the types in this file?

**Comment by BSalita** [comment]
Could you list the remaining controversial parts? I'm ok with another round of justification. I believe you're thinking some of these changes are optional because you're not seeing the compiler warnings that I'm seeing. The compiler warnings are largely a result of the existing code not dealing with 32/64 issues properly.

File: src/passes/Print.cpp (hunk 10 above)
Comment (BSalita) [inline src/passes/Print.cpp]: Not only would we have to change to IndexedVector but also to IndexedMap.

File: src/wasm-binary.h (hunk 11 above)
Comment (BSalita) [inline src/wasm-binary.h]: Since `SIZE32LEB` performs a useful task, properly casting `size_t` to 32-bit, we should do the opposite. Accept it for now. Perhaps someone can come up with a better solution for `SIZE32LEB` specifically or all of the `*32LEB` cast in general.

File: src/passes/Print.cpp (hunk 10 above)
Comment (kripken) [inline src/passes/Print.cpp]: Yes, those two are the relevant types.

File: src/wasm-binary.h (hunk 11 above)
Comment (kripken) [inline src/wasm-binary.h]: That is useful, but I think the real question is where a non-32-bit value comes from. It's an array or map, perhaps those types should return 32-bit values anyhow, as mentioned before.

**Comment by kripken** [comment]
I am not convinced those warnings are serious enough to justify clutter in the code.

If they are serious, let's try to find a way to fix them that doesn't add clutter (but that might be a longer-term thing).

But are they? If the risk is trying to truncate a 64-bit value that doesn't fit in 32 bits, we should do it on external inputs, but on internal data that is assumed valid, I don't think the clutter is worth it. Readability matters more.

**Comment by BSalita** [comment]
Regarding warnings, my belief is that you treat them as errors. That's clearly not shared by others on this forum. My own code is warning-less is all environments. There are some exceptions such as warnings due to bugs in compilers (non-existent now), or serving as To Do reminders that should be addressed later.

Regarding clutter, if you mean syntax needed to suppress warnings due to type differences (casting), I don't see that as clutter. I see it as necessary. When you have source code that needs to run on multiple compilers, different endians, different architectures (32 vs 64, arm vs Intel), different OSs, incorporate legacy considerations, optimal performance, differing loaders/linkers, signed/unsigned compiler diffs, I don't know how you do that other than adding syntax to the code at issue. To be clear, clutter is in your future and you won't be able to prevent it. Of course if you want to limit the source code to a couple compilers, and just 32 or 64-bit, you'll have no more clutter than you have now.

There are ways to mitigate clutter such as your suggestion to create new vector and map classes just to workaround size() casting. A way to hide endian clutter is to create endian template classes with operator overloads. However, all these has trade-offs. I fear such classes will add complexity, add CPU cycles, and redirect development time better spent elsewhere.

File: src/wasm-binary.h (hunk 11 above)
Comment (dschuff) [inline src/wasm-binary.h]: What about just changing the constructor of LEB to take a `size_t` argument instead of a `T`? That would force all arguments to be implicitly promoted at the construction site when necessary, and then the implementation would cast down to `T` and could also assert that it doesn't overflow.

File: src/emscripten-optimizer/simple_ast.h (hunk 2 above)
Comment (dschuff) [inline src/emscripten-optimizer/simple_ast.h]: Can we actually do what I suggested above (i.e. change `(char)('0' + (num / 10));` to `'0' + (char)(num / 10);`?

File: src/passes/Print.cpp (hunk 10 above)
Comment (dschuff) [inline src/passes/Print.cpp]: What about something like #537 instead?

**Comment by dschuff** [comment]
I agree that we want to treat warnings as errors. I don't necessarily agree that all warnings on all compilers for all platforms must be enabled. I do think that for the most part it's worth getting better type hygiene, and I also think that if we find ourselves doing a lot of static or c-style casting, then maybe we should look for other ways.

File: src/emscripten-optimizer/simple_ast.h (hunk 2 above)
Comment (BSalita) [inline src/emscripten-optimizer/simple_ast.h]: I've not wanted to create more changes to this PR. However, since you asked twice, I went ahead and coded per your request.

**Comment by dschuff** [comment]
OK after looking some more, I think we should:
1. Remove `toIndex` from this PR
2. Remove `SIZE32LEB` from this PR.
3. Change the constructor of `LEB` to take a `size_t` argument as I suggested [above](https://github.com/WebAssembly/binaryen/pull/491#discussion_r64194272) (that can be in this PR because it's a pretty trivial change).
4. Land #537 or something like it in a separate PR.

If you do 1 and 2 and optionally 3 for this PR, then it LGTM.

**Comment by BSalita** [comment]
End of the day here. I'll tackle tomorrow. 

    On Monday, May 23, 2016 9:09 PM, Derek Schuff <notifications@github.com> wrote:
 

 OK after looking some more, I think we should:
1. Remove toIndex from this PR
2. Remove SIZE32LEB from this PR.
3. Change the constructor of LEB to take a size_t argument as I suggested above (that can be in this PR because it's a pretty trivial change).
4. Land #537 or something like it in a separate PR.If you do 1 and 2 and optionally 3 for this PR, then it LGTM.—

**Comment by kripken** [comment]
This has been inactive for a long time, is it still relevant?

**Comment by dschuff** [comment]
It could be if we wanted. In eafc06f8cf2b815d151b67c67086f33b3dae6b8b I disabled the warnings that create all the noise about this. But it still seems like it might be valuable to fix these type issues.

**Comment by dschuff** [comment]
On the other hand, several of the codebases I work with (e.g. LLVM and Chromium) also disable these warnings, so it may not be worth fixing all of these.

**Comment by kripken** [comment]
Ok. Let's close this for now then with those warnings disabled, we can revisit the general issue later if there is interest.
